### PR TITLE
feat: reorder kanban columns

### LINKED ISSUE #1641: Implement possibilty to move Kaban colomn
### Scope
- Opportunities (kanban)

At present, users do not have the ability to change the position of columns on the Kanban. 

# Desired Design:

We want to add a column reordering functionality. This can be achieved by implementing the following changes:
When users click on the name column within the BoardColumnEditTitleMenu, an options to "Move Left" and "Move Right" is present.
Selecting "Move Left" the colomn should shift to the left, and selecting "Move Right" should shift it to the right.

https://github.com/twentyhq/twenty/assets/102394218/24405477-cb6e-4c0e-a911-4c1f95367dca

https://www.figma.com/file/xt8O9mFeLl46C5InWwoMrN/Twenty?type=design&node-id=11015-55936&mode=design&t=9eeUr4AoZTMi9TEI-4

### PR DESCRIPTION
**Pull request by AdityaPimpalkar**
closes: #1641

https://github.com/twentyhq/twenty/assets/13139771/2956d581-844d-41ed-a7a4-62fc8ad8e0dc

### CODE REVIEW

File: front/src/modules/ui/board/components/EntityBoardColumn.tsx [hunk 1]
```diff
@@ -78,10 +86,12 @@ export const EntityBoardColumn = ({
...
+          column={column}
           title={column.title}
           color={column.colorCode}
           totalAmount={boardColumnTotal}
-          isFirstColumn={column.index === 0}
+          isFirstColumn={isFirstColumn}
+          isLastColumn={isLastColumn}
           numChildren={cardIds.length}
           stageId={column.id}
         >
```
Comment (AdityaPimpalkar) [inline front/src/modules/ui/board/components/EntityBoardColumn.tsx:97]: Prop drilling is getting a bit out of hand here, should we pass some of these into a context?

**Review by AdityaPimpalkar — COMMENTED** [review]
Need some suggestions, thanks! :)

File: front/src/modules/ui/board/components/EntityBoardColumn.tsx (hunk 1 above)
Comment (charlesBochet) [reply front/src/modules/ui/board/components/EntityBoardColumn.tsx:97]: Yes! you can create a BoardColumnContext and store everything there

File: front/src/modules/ui/board/hooks/useBoardColumns.ts [hunk 2]
```diff
@@ -0,0 +1,49 @@
+import { useRecoilState } from 'recoil';
+
+import { boardColumnsState } from '../states/boardColumnsState';
+import { BoardColumnDefinition } from '../types/BoardColumnDefinition';
+
+export const useBoardColumnns = () => {
+  const [boardColumns, setBoardColumns] = useRecoilState(boardColumnsState);
```
[the remark this answers is not in the document]
Comment (charlesBochet) [reply front/src/modules/ui/board/hooks/useBoardColumns.ts:7]: I don't think there is a bug in recoil. It's probably in our code that we are not reacting to recoil state change.
What's the issue?  I can help investigate!

**Review by charlesBochet — COMMENTED** [review]
Another future masterpiece, thank. you @AdityaPimpalkar

File: front/src/modules/ui/board/hooks/useBoardColumns.ts [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+
+  const handleColumnMove = (
+    direction: string,
+    column: BoardColumnDefinition,
+  ) => {
+    const currentColumnArrayIndex = boardColumns.findIndex(
+      (tableColumn) => tableColumn.id === column.id,
+    );
+
+    const targetColumnArrayIndex =
```
Comment (charlesBochet) [inline front/src/modules/ui/board/hooks/useBoardColumns.ts:17]: this looks like duplicated logic with Table. Can we extract it to a common hook? useMoveViewFields() you can put it into the views folder

File: front/src/modules/ui/board/components/EntityBoardColumn.tsx (hunk 1 above)
Comment (AdityaPimpalkar) [reply front/src/modules/ui/board/components/EntityBoardColumn.tsx]: Done!

File: front/src/modules/ui/board/hooks/useBoardColumns.ts (hunk 3 above)
Comment (AdityaPimpalkar) [reply front/src/modules/ui/board/hooks/useBoardColumns.ts]: Done!

File: front/src/modules/ui/board/hooks/useBoardColumns.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (AdityaPimpalkar) [reply front/src/modules/ui/board/hooks/useBoardColumns.ts:9]: The `ViewBar` in `BoardHeader` is using `canPersistViewFields: canPersistBoardCardFields` meaning the view bar visibility is only triggered only when view fields visibility is changing. I think we may need to add another state check just like board card fields for board columns too?

However, this still didn't answer why the state change is not visible, still looking into it.

**Comment by AdityaPimpalkar** [comment]
`HooksCompanyBoardEffect` was being re-rendered on state change and that hook makes a new db call on each render so my recoil state change was being replaced every time. 

I found a fix, but I could use your input to make it better. Thanks :)

https://github.com/twentyhq/twenty/assets/13139771/8a34b488-f122-4d9d-aa13-c2af8e3bc516

File: front/src/modules/ui/board/hooks/useBoardColumns.ts (hunk 3 above)
Comment (thaisguigon) [reply front/src/modules/ui/board/hooks/useBoardColumns.ts]: Are we sure we want to name this `useMoveViewFields`? Board columns are PipelineStages, not ViewFields, unlike Table columns. Couldn't this cause confusion in the future @charlesBochet? We could perhaps name this `useMoveViewColumns` for instance?

File: front/src/modules/views/hooks/useMoveViewFields.ts [hunk 4]
```diff
@@ -0,0 +1,29 @@
+export const useMoveViewFields = () => {
+  const handleColumnMove = <T extends { index: number }>(
+    direction: 'left' | 'right',
+    currentArrayindex: number,
+    targetArray: T[],
+  ) => {
+    const targetArrayIndex =
+      direction === 'left' ? currentArrayindex - 1 : currentArrayindex + 1;
+    if (currentArrayindex >= 0 && targetArrayIndex >= 0) {
```
[the remark this answers is not in the document]
Comment (AdityaPimpalkar) [reply front/src/modules/views/hooks/useMoveViewFields.ts]: I agree! Thanks :)

File: front/src/modules/views/hooks/useBoardViewFields.ts [hunk 5]
```diff
@@ -103,6 +110,23 @@ export const useBoardViewFields = ({
     );
   };
 
+  const updatedPipelineStages = (stages: BoardColumnDefinition[]) => {
```
[the remark this answers is not in the document]
Comment (AdityaPimpalkar) [reply front/src/modules/views/hooks/useBoardViewFields.ts]: We already have a `useBoardColumns`, moved the function there 👍

File: front/src/modules/views/hooks/useBoardViewFields.ts [hunk 6]
```diff
@@ -165,5 +189,11 @@ export const useBoardViewFields = ({
     return refetch();
   };
 
-  return { createViewFields, persistCardFields };
+  const persistBoardColumns = async () => {
+    if (!currentViewId) return;
+    setSavedBoardColumns(boardColumns);
```
[the remark this answers is not in the document]
Comment (AdityaPimpalkar) [reply front/src/modules/views/hooks/useBoardViewFields.ts]: Done! :)

File: front/src/modules/ui/board/hooks/useBoardColumns.ts (hunk 3 above)
Comment (charlesBochet) [reply front/src/modules/ui/board/hooks/useBoardColumns.ts]: It's a good point!

**Review by charlesBochet — APPROVED** [review]
Thank you!

**Pull request merged**